tools/adios2pio-nm/adios2pio-nm-lib.cxx

```diff
@@ -1525,229 +1569,337 @@ int ConvertVariableTimedPutVar(IOVector &bpIO, EngineVector &bpReader,
+        for (size_t i = 0; i < db_blocks.size(); i++)
+        {
+            decompid_var.SetBlockSelection(i);
+            bpReader[0].Get(decompid_var, decomp_buffer + tmp_idx, adios2::Mode::Sync);
+            tmp_idx += db_blocks[i].Count[0];
+        }
 
-    /* Different decompositions at different frames */
-    char decomp_varname[PIO_MAX_NAME];
-    char frame_varname[PIO_MAX_NAME];
-    char fillval_varname[PIO_MAX_NAME];
-    char decompname[PIO_MAX_NAME];
-    snprintf(decomp_varname, PIO_MAX_NAME, "decomp_id/%s", varname.c_str());
-    snprintf(frame_varname, PIO_MAX_NAME, "frame_id/%s", varname.c_str());
-    snprintf(fillval_varname, PIO_MAX_NAME, "fillval_id/%s", varname.c_str());
-    int decomp_id, frame_id, fillval_exist;
-    char fillval_id[PIO_MAX_NAME];
-
-    // TAHSIN -- THIS IS GETTING CONFUSING. NEED TO THINK ABOUT time steps.
-    for (; ts < nsteps; ++ts)
+        /* read fillval_id buffer */
+        adios2::Variable<T> fillval_var = bpIO[0].InquireVariable<T>("/__pio__/track/fillval_id/" + variable_name);
+        if (fillval_var)
+        {
+            const auto fb_blocks = bpReader[0].BlocksInfo(fillval_var, time_step);
+            fillval_buffer = (char*)calloc(nsteps, sizeof(T));
+            tmp_idx = 0;
+            std::vector<T> fb_tmp;
+            int fb_tmp_size = 0;
+            for (size_t i = 0; i < fb_blocks.size(); i++)
+            {
+                fillval_var.SetBlockSelection(i);
+                bpReader[0].Get(fillval_var, fb_tmp, adios2::Mode::Sync);
+                fb_tmp_size = fb_tmp.size() * sizeof(T);
+                memcpy(fillval_buffer + tmp_idx, fb_tmp.data(), fb_tmp_size);
+                tmp_idx += fb_tmp_size;
+            }
+        }
+    }
+    catch (const std::exception &e)
+    {
+        std::cout << "TIME STEPS Error: " << e.what() << std::endl;
+        return BP2PIO_ERROR;
+    }
+    catch (...)
+    {
+        return BP2PIO_ERROR;
+    }
+
+    /* Find block locations for each writer in each block for all time steps */
+    std::vector<std::vector<int> > writer_block_id;
+    int num_procs = 0;
+    for (size_t i = 0; i < block_procs.size(); i++)
+    {
+        num_procs += block_procs[i];
+    }
+    writer_block_id.resize(num_procs);
+    for (int i = 0; i < num_procs; i++)
+    {
+        writer_block_id[i].resize(nsteps);
+        for (int j = 0; j < nsteps; j++)
+        {
+            writer_block_id[i][j] = -1; /* nothing written out by proc i at time step j */
+        }
+    }
+    adios2::Variable<int> blk_var = bpIO[0].InquireVariable<int>("/__pio__/track/num_data_block_writers/" + variable_name);
+    const auto blk_blocks = bpReader[0].BlocksInfo(blk_var, time_step);
+    int num_bp_blocks_per_group = blk_blocks.size() / block_list.size();
+    if ((num_bp_blocks_per_group*block_list.size()) != blk_blocks.size())
+    {
+        printf("ERROR: #blocks: %lu !=  #written: %lu\n", num_bp_blocks_per_group * block_list.size(), blk_blocks.size());
```

Comment:
std:cerr instead of printf ?

Reply:
We can use fprintf(stderr, ...) so far, which is also used in a few other places in this file.